result.nim

```diff
@@ -0,0 +1,352 @@
+    ##
+    ## # Other implemenations in nim
+    ##
+    ## There are other implementations in nim that you might prefer:
+    ## * Either from nimfp: https://github.com/vegansk/nimfp/blob/master/src/fp/either.nim
+    ## * result_type: https://github.com/kapralos/result_type/
+    ##
+    ## # Implementation notes
+    ##
+    ## This implementation is mostly based on the one in rust. Compared to it,
+    ## there are a few differences - if know of creative ways to improve things,
+    ## I'm all ears.
+    ##
+    ## * Rust has the enum variants which lend themselves to nice construction
+    ##   where the full Result type isn't needed: `Err("some error")` doesn't
+    ##   need to know value type - maybe some creative converter or something
+    ##   can deal with this?
+    ## * Nim templates allow us to fail fast without extra effort, meaning the
+    ##   other side of `and`/`or` isn't evaluated unless necessary - nice!
+    ## * In Nim, we have exceptions - when using this library, we'll raise the
+    ##   standard crop of Nim errors when trying to access the error of a value
+    ##   and vice versa - this fits better with Nim but costs space and
+    ##   performance - need to think about this - rust will simply panic, and
+    ##   everyone seems more or less happy with that..
+    ## * Rust uses From traits to deal with result translation as the result
+    ##   travels up the call stack - needs more tinkering - some implicit
+    ##   conversions would be nice here
+
+    case isOk*: bool
+    of false:
+      error*: E
+    of true:
+      value*: T
+
+  # TODO cannot raise generic error, so cannot include error value :(
+  # https://github.com/nim-lang/Nim/issues/7845
+  ResultError = object of Exception
+
+# TODO check that r is a Result
+proc ok*(r: typedesc, v: auto): r =
+  ## Initialize a result with a success and value
+  ## Example: `Result[int].ok(42)`
+  r(isOk: true, value: v)
+proc ok*[T, E](self: var Result[T, E], v: auto) =
+  ## Set the result to success and update value
+  ## Example: `result.ok(42)`
+  self = Result[T, E].ok(v)
+
+proc err*(r: typedesc, e: auto): r =
+  ## Initialize the result to an error
+  ## Example: `Result[int].err("uh-oh")`
+  r(isOk: false, error: e)
+
+proc err*[T, E](self: var Result[T, E], v: auto) =
+  ## Set the result as an error
+  ## Example: `result.err("uh-oh")`
+  self = Result[T, E].err(v)
+
+template isErr*[T, E](self: Result[T, E]): bool = not self.isOk
+
+proc map*[T, E, A](self: Result[T, E], f: proc(x: T): A {.closure.}): Result[A, E] =
+  ## Transform value using f, or return error
+  if self.isOk: result.ok(f(self.value))
+  else: result.err(self.error)
+
+proc mapErr*[T, E, A](self: Result[T, E], f: proc(x: E): A {.closure.}): Result[T, A] =
+  ## Transform error using f, or return value
+  if self.isOk: result.ok(self.value)
+  else: result.err(f(self.error))
+
+proc mapConvert*[T0, E0](self: Result[T0, E0], T1: typedesc): Result[T1, E0] =
```

Comment:
There are some filed issues in Nim when using generic converters. Once fixed, perhaps it will be possible to automatically lift the conversion here.

Reply:
yeah, I thought briefly about a converter, but you'd perhaps also need some type deduction magic based on the left side of the assignment. 

this could use improvement for sure, as a common use case is to convert error types between api layers. rust has the `from` trait for this purpose: https://doc.rust-lang.org/1.22.0/std/convert/trait.From.html